### ISSUE TITLE
dart2js: deferred parts contain repeated type rules

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#41980]
I noticed that small deferred loaded parts all contain the same set of duplicate rules:
```js
  H._Universe_addRules(init.typeUniverse, JSON.parse('{"AnimationPlaybackEvent":"Event2","AnalyserNode":"AudioNode","ConstantSourceNode":"AudioScheduledSourceNode","AudioContext":"BaseAudioContext","AnimateElement":"SvgElement","AnimationElement":"SvgElement","TSpanElement":"TextContentElement","TextPositioningElement":"TextContentElement","ClipPathElement":"GraphicsElement","LinearGradientElement":"_GradientElement","CircleElement":"GeometryElement","OpenDBRequest":"Request1","BatteryManager":"EventTarget","_ResourceProgressEvent":"ProgressEvent","BRElement":"HtmlElement","_DocumentType":"Node0","XmlDocument":"Document","PerformanceLongTaskTiming":"PerformanceEntry","AbortPaymentEvent":"ExtendableEvent","ShadowRoot":"DocumentFragment","DedicatedWorkerGlobalScope":"WorkerGlobalScope","MidiInput":"MidiPort","AbsoluteOrientationSensor":"Sensor","OrientationSensor":"Sensor","HttpRequestUpload":"HttpRequestEventTarget","HtmlFormControlsCollection":"HtmlCollection","CssCharsetRule":"CssRule","CssStyleSheet":"StyleSheet","Comment":"CharacterData","CanvasCaptureMediaStreamTrack":"MediaStreamTrack","CDataSection":"Text","BackgroundFetchFailEvent":"BackgroundFetchEvent","AudioElement":"MediaElement"}'));
```
Obviously, these rules should not be redefined by each part.

/cc @fishythefish

**Comment by fishythefish** [comment]
I don't think we ever check that each class is added to the global ruleset only once. Each fragment adds all the classes (that need rules) from its libraries. If each class belongs to only one fragment, then there should be no duplication, but it looks like a bunch of `dart:html` classes are being attributed to multiple fragments (possibly because the fragments all point to the `dart:html` library).

If a library is shared by multiple fragments, whichever output unit is loaded first will need to inject the rules, but I don't think there's always a canonical choice for which fragment to pick.
